### PR TITLE
Issue 12677 role selector workflow step editor

### PR DESCRIPTION
**Pull request by fabrizzio-dotCMS**

### CODE REVIEW

File: dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/business/WorkflowAPITest.java [hunk 1]
```diff
@@ -761,6 +761,13 @@ public void findActions() throws DotDataException, DotSecurityException {
         //check valid action for
         actions = workflowAPI.findActions(steps, reviewerUser);
         assertTrue(null != actions && actions.size() == 2);
+
+        actions = workflowAPI.findActions(workflowScheme1Step2, reviewer);
+        assertTrue(null != actions && actions.size() == 1);
```
Comment (jdotcms) [inline dotCMS/src/integration-test/java/com/dotmarketing/portlets/workflows/business/WorkflowAPITest.java]: this kind of things is better to separated on two assert
assertNotNull
assertEquals (2, actions.size);

Give you more info about what is the error

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/ajax/WfRoleStoreAjax.java [hunk 2]
```diff
@@ -175,15 +176,15 @@ public void assignable ( HttpServletRequest request, HttpServletResponse respons
...
+            response.getWriter().write( rolesToJson( roleList, includeFake, includeFake ) );
         } catch ( Exception e ) {
             Logger.error( WfRoleStoreAjax.class, e.getMessage(), e );
         }
 
 
     }
 
-    private String rolesToJson ( List<Role> roles, boolean includeFake ) throws IOException, DotDataException, LanguageException {
+    private String rolesToJson ( List<Role> roles, boolean includeFake, boolean includeLabelAll ) throws IOException, DotDataException, LanguageException {
```
[the remark this answers is not in the document]
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/portlets/workflows/ajax/WfRoleStoreAjax.java]: it is a legacy tradition to include flag parameters as a boolean in the functions, we are trying to leave it and start doing things such as

allRolesToJson or something like that.

In that case I think it is not an issue b/c it is a private method but just to keep in mind

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/ajax/WfRoleStoreAjax.java [hunk 3]
```diff
@@ -194,12 +195,19 @@ private String rolesToJson ( List<Role> roles, boolean includeFake ) throws IOEx
         Map<String, Object> map = null;
 
         if(includeFake) {
-        	map = new HashMap<String, Object>();
+        	map = new HashMap<>();
 	        map.put( "name", "" );
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/workflows/ajax/WfRoleStoreAjax.java]: replace "" by StringPool.EMPTY

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPI.java [hunk 4]
```diff
@@ -340,6 +350,18 @@ public void reorderAction(final WorkflowAction action,
 	public List<WorkflowAction> findActions(WorkflowStep step, User user, Permissionable permissionable) throws DotDataException,
 			DotSecurityException;
 
+
+	/**
+	 *
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPI.java]: Add documentation here

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java [hunk 5]
```diff
@@ -819,6 +819,17 @@ public void removeAttachedFile(WorkflowTask task, String fileInode) throws DotDa
         return filterActionsCollection(actions, user, true, permissionable);
     }
 
+	public List<WorkflowAction> findActions(WorkflowStep step, Role role) throws DotDataException,
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java]: set always params to final

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java [hunk 6]
```diff
@@ -819,6 +819,17 @@ public void removeAttachedFile(WorkflowTask task, String fileInode) throws DotDa
         return filterActionsCollection(actions, user, true, permissionable);
     }
 
+	public List<WorkflowAction> findActions(WorkflowStep step, Role role) throws DotDataException,
+           DotSecurityException {
+
+		if(null == step) {
+			return Collections.emptyList();
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java]: I am wondering if we should start using Optional for this scenarios

Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java]: or even throw an Illegal Arg Exception

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java (hunk 5 above)
Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java]: Remember to add the @closeDbIfOpened annotation

Comment (jdotcms) [reply dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java]: and the override too

File: dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java [hunk 7]
```diff
@@ -2082,6 +2093,33 @@ public void updateStepReferences(final String stepId, final String replacementSt
...
+			final Role role) throws DotDataException {
+
+		final List<WorkflowAction> permissionables = new ArrayList<>(actions);
+		if (permissionables.isEmpty()) {
+			return permissionables;
+		}
+
+		final Iterator<WorkflowAction> iterator = permissionables.iterator();
+		while (iterator.hasNext()) {
+			final WorkflowAction wa = iterator.next();
```
Comment (jdotcms) [inline dotCMS/src/main/java/com/dotmarketing/portlets/workflows/business/WorkflowAPIImpl.java]: we are trying to use meaningful names, you can rename to workflowAction

**Review by jgambarios — CHANGES_REQUESTED** [review]
Great work Fabrizzio!, now that you are there I think we can remove these extra queries that I think no one is using: https://github.com/dotCMS/core/blob/issue-12677-role-selector-workflow-step-editor/dotCMS/src/main/webapp/html/portlet/ext/workflows/schemes/view_steps.jsp#L21-L26

**Review by jdotcms — APPROVED** [review]

**Comment by fabrizzio-dotCMS** [comment]
Again the requested changes were made. The unused queries are no longer there.

**Review by jgambarios — APPROVED** [review]

**Pull request merged**